File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.count = 0
        self.ans = 0
    def sumNodes(self, root: Optional[TreeNode]) ->int:
        if(root is None):
            return 0
        
        self.count += 1
        left = self.sumNodes(root.left)
        right = self.sumNodes(root.right)
        
        return root.val + left + right
    
    def averageNodes(self, root: Optional[TreeNode]):
        if(root is None):
            return
        
        self.count = 0
        avg = self.sumNodes(root)
        avg = avg//self.count
        if(avg == root.val):
            self.ans += 1
        
        self.averageNodes(root.left)
        self.averageNodes(root.right)
        
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        if(root.left is None and root.right is None):
            return 1
        
        
        self.averageNodes(root)
        return self.ans
        
        

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

int ans = 0;

class Solution 
{
public:
    int sum(TreeNode* root, int& count)
    {
        if(root == nullptr)
            return 0;
        
        count++;
        int left = sum(root->left, count);
        int right= sum(root->right,count);
        
        return (root->val + left + right);
    }
    
    void average(TreeNode* root)
    {
        if(root == nullptr)
            return;
        int count = 0;
        int avg = sum(root, count);
        avg /= count;
        if(avg == root->val)
            ans++;
        
        average(root->left);
        average(root->right);
    }
    
    int averageOfSubtree(TreeNode* root) 
    {
        ans = 0;
        average(root);
        return ans;
    }
};